### PR TITLE
chore: implement new key semantics in queries

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

This change implements the change in key semantics in queries outlined in [KLIP-24](#5115).

With this change the projection of persistent queries must include any key columns. For everything other than a `GROUP BY` this will be a single column. `GROUP BY` supports multiple key columns, (even though they end up munged into a single string key at the moment).

To detect the missing key columns I've extended `LogicalPlanner` and the logical `PlanNode`s. When building a persistent query the planner calls `validateKeyPresent` on the final node in the logical plan before the output node: this will either be a `AggregateNode` or a `FinalProjectNode`, depending on where there is a `GROUP BY` clause.  

The `validateKeyPresent` method percolate up the logical plan from child to parent nodes until it reaches a node that knows how to handle it.

* for any query with `GROUP BY`, this will be the `AggregateNode`.
* from any query with `PARTITION BY`, this will be the `RepartitionNode`.
* for join queries, with no aggregation or repartition, this will be the `JoinNode`.
* for simple queries, (no join, repartition or aggregation), this will be the `DataSourceNode`.

Joins are the complicated one, as they can have a synthetic key column. This necessitated adding an optional `keyCol` to the schema of joins, which defaults to the legacy `ROWKEY` name.

To get an error message that makes sense to the user required that original, as well as rewritten, expressions got passed around.  This is a bit ugly, but refactoring to remove this is out of scope.

There are some remaining tasks to do to complete this work and tidy things up - but they will be done in follow up tasks.  Also a lot of the added complexity will be removed once the 'any key name` feature is turned on, and the legacy code removed.

### Reviewing:

This is a big change. Apologies for this. Not sure how I can could break this down really.

* I'd start with looking at the changes to the josn QTT tests to get a feel for what's changed in the syntax.
* Then look at the logical plan and its nodes
* then check out the changes to `StreamStreamJoin` and the table variants.

I'll add some notes to the PR to hopefully help.

### Testing done 

Usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-common/src/main/java/io/confluent/ksql/util/GrammaticalJoiner.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * An object which joins pieces of text, with a separator and additional final word, allowing more
+ * grammatically correct joining, e,g. "a, b, c or d".
+ */
+public final class GrammaticalJoiner {
```
Comment (big-andy-coates) [inline ksqldb-common/src/main/java/io/confluent/ksql/util/GrammaticalJoiner.java:28]: Added have more human friendly phrasing of error messages.  If anyone knows of an existing utility to do this I'm happy to switch...

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java [hunk 2]
```diff
@@ -149,6 +158,11 @@ private OutputNode buildOutputNode(final PlanNode sourcePlanNode) {
 
     final Into intoDataSource = analysis.getInto().get();
 
+    if (ksqlConfig.getBoolean(KsqlConfig.KSQL_ANY_KEY_NAME_ENABLED)) {
+      ((VerifiableNode) sourcePlanNode)
+          .validateKeyPresent(intoDataSource.getName());
+    }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:164]: This is the call to throw, should the projection not include the key columns.

Once the `any key name` functionality has been turned on, this check can move into `KsqlStructuredDataOutputNode` to make it more OO.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java [hunk 3]
```diff
@@ -523,16 +598,75 @@ private PlanNode buildJoin(final Join root, final String prefix) {
...
+        joinKey.rewriteWith(refRewriter::process),
+        finalJoin,
         left,
         right,
         root.getInfo().getWithinExpression()
     );
   }
 
-  private DataSourceNode buildNonJoinNode(final AliasedDataSource dataSource) {
+  private JoinKey buildJoinKey(final Join join) {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:616]: To be able to determine the schema of the `JoinNode` we must first know which of the join expressions, if any, will be used as the key column in the result.   This is a bit of a chicken and egg problem, and this isn't a great solution.

I have thoughts on how this can be resolved in a follow up PR.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/Projection.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import io.confluent.ksql.parser.tree.SingleColumn;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Holds information about in a projection
+ */
+@Immutable
+public final class Projection {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/Projection.java:35]: Added this class to help with working out what's in the projection without having to compile all the `SelectExpression`s, mainly as the compilation requires the source schema, as we need to work out what's in the projection of joins to work out what the schema is!

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/DataSourceNode.java [hunk 5]
```diff
@@ -139,28 +153,6 @@ private static LogicalSchema buildSchema(final DataSource dataSource) {
         .withPseudoAndKeyColsInValue(dataSource.getKsqlTopic().getKeyFormat().isWindowed());
   }
 
-  @SuppressWarnings("UnstableApiUsage")
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/DataSourceNode.java:142]: Moved to parent class.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 6]
```diff
@@ -112,15 +145,59 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
...
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  @Override
+  void validateKeyPresent(final SourceName sinkName, final Projection projection) {
+
+    final boolean atLeastOneKey = Streams.concat(
+        joinKey.getAvailableKeys().stream(),
+        joinKey.getOriginalAvailableKeys().stream()
+    ).anyMatch(projection::containsExpression);
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:179]: Both are unfortunately required at the moment because of the mess around rewritten vs original expressions.

The `projection` passed in knows about rewritten column references e.g. `T_ID` , not `T.ID`. So we need to use rewritten columns returned by `joinKey.getAvailableKeys()`.  

Unfortunately, the `T_ID` won't match a `T.*` in the projection. For that to match we need the original `T.ID` - hence we need both for now.

I've got a todo locally to see if this can be avoided. Fixing it will likely mean rethinking how the rewrites are done.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json [hunk 7]
```diff
@@ -583,29 +834,65 @@
...
+      "inputs": [
+        {"topic": "right2", "key": 0, "value": {"V0": 3}, "timestamp": 10},
+        {"topic": "right", "key": 0, "value": {"V0": 2}, "timestamp": 11},
+        {"topic": "left", "key": 0, "value": {"V0": 1}, "timestamp": 12},
+        {"topic": "right", "key": 0, "value": {"V0": 4}, "timestamp": 13},
+        {"topic": "left", "key": 0, "value": {"V0": 5}, "timestamp": 100000},
+        {"topic": "right", "key": 0, "value": {"V0": 6}, "timestamp": 100001}
+      ],
+      "outputs": [
+        {"topic": "_confluent-ksql-some.ksql.service.idquery_CSAS_OUTPUT_0-Join-repartition", "key": null, "value": {"S1_V0": null, "S1_ROWTIME": null, "S1_ID": null, "S2_V0": 2, "S2_ROWTIME": 11, "S2_ID": 0, "KSQL_COL_0": null}, "timestamp": 11},
```
Comment (big-andy-coates) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json:873]: Note: the synthetic key column from the full outer join `KSQL_COL_0` is included in the value of the repartition topic, but is always `null`.   This is not pretty and I'd like to find a solution to this. But this may be better fixed in a separate PR.

Created https://github.com/confluentinc/ksql/issues/5258 to track.

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/KsqlValueJoiner.java [hunk 8]
```diff
@@ -15,38 +15,50 @@
...
       row.appendAll(right.values());
     } else {
-      fillWithNulls(row, rightSchema.value().size());
+      fillWithNulls(row, rightCount);
     }
 
+    // Potentially append additional nulls as a holder for a synthetic key columns.
+    // These columns are not populated, as they are not accessed, but must be present for the row
+    // to match the rows schema.
+    fillWithNulls(row, additionalCount);
```
Comment (big-andy-coates) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/KsqlValueJoiner.java:60]: Again, not nice - tracked by https://github.com/confluentinc/ksql/issues/5258 as well.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json [hunk 9]
```diff
@@ -196,82 +235,245 @@
...
+      "statements": [
+        "CREATE STREAM TEST (ID INT KEY, data STRING) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT AS SELECT DATA AS COUNT, COUNT(*) AS COUNT FROM TEST GROUP BY DATA;"
+      ],
+      "properties": {
+        "ksql.any.key.name.enabled": true
+      },
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Duplicate values found in schema: `COUNT` BIGINT"
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:469]: Hummm... this error comes from `LogicalSchema` and is specifically about duplicate value columns, as opposed to key columns.

Updated to "Duplicate value columns found in schema".

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/joinkey.json [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+          {"name": "OUTPUT", "type": "table", "schema": "ID INT KEY, L_ID INT, R_ID INT, L_V0 INT, R_V1 INT"}
+        ]
+      }
+    },
+    {
+      "name": "multi-join",
+      "statements": [
+        "CREATE TABLE L (ID INT PRIMARY KEY, V0 INT, V1 INT) WITH (kafka_topic='left_topic', value_format='JSON');",
+        "CREATE TABLE R (ID INT PRIMARY KEY, V0 INT, V1 INT) WITH (kafka_topic='right_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT as SELECT JOINKEY(L.ID, R.ID) AS ID, L.ID, R.ID, L.V0, R.V1 FROM L FULL OUTER JOIN R ON L.id = R.id;"
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/joinkey.json:85]: The name would suggest so ;)

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java [hunk 11]
```diff
@@ -523,16 +598,75 @@ private PlanNode buildJoin(final Join root, final String prefix) {
...
+    if (info.getType() == JoinType.OUTER) {
+      return JoinKey.syntheticColumn(info.getLeftJoinExpression(), info.getRightJoinExpression());
+    }
+
+    final boolean leftIsCol = info.getLeftJoinExpression() instanceof QualifiedColumnReferenceExp;
+    final boolean rightIsCol = info.getRightJoinExpression() instanceof QualifiedColumnReferenceExp;
+
+    // - neither join expression is a column reference: if neither is a column reference, by
+    //   definition the column that holds the result is not equal to a column in either source: it
+    //   is a _new_ column.
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:637]: Good catch.

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/RewrittenAnalysis.java [hunk 12]
```diff
@@ -59,6 +59,10 @@ public RewrittenAnalysis(
     this.rewriter = Objects.requireNonNull(rewriter, "rewriter");
   }
 
+  public ImmutableAnalysis original() {
+    return original;
+  }
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/RewrittenAnalysis.java:64]: I know, I don't like putting it back either.

The originals are needed for error messages, so that we get errors about `T.ID` not being in the projection, (which is what the user needs to add), and not `T_ID` which would result in an error if they added it.

Improving the way this handled is probably best left outside the scope of this PR though. :'(

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java [hunk 13]
```diff
@@ -554,17 +688,55 @@ private DataSourceNode buildNonJoinNode(final AliasedDataSource dataSource) {
...
+
+    for (final SelectExpression select : projection) {
+      final Expression expression = select.getExpression();
+
+      final SqlType expressionType = expressionTypeManager
+          .getExpressionSqlType(expression);
+
+      final boolean keyColumn = expression instanceof ColumnReferenceExp
+          ? parentSchema.isKeyColumn(((ColumnReferenceExp) expression).getColumnName())
+          : parentSchema.isKeyColumn(select.getAlias());
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:715]: You are right - updated to:

```java
final boolean keyColumn = expression instanceof ColumnReferenceExp 
          && parentSchema.isKeyColumn(((ColumnReferenceExp) expression).getColumnName());
```

(Really appreciate your thorough review! My heads been in this code for too long. Can't see the wood for the trees...)

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java [hunk 14]
```diff
@@ -589,54 +761,101 @@ private LogicalSchema buildAggregateSchema(
...
-            .orElseGet(keyColNameGen::nextKsqlColAlias);
+
+        keyName = Optional.of(groupBy.getAlias()
+            .orElseGet(() -> ColumnNames.columnAliasGenerator(Stream.of(
+                sourceSchema,
+                LogicalSchema.builder()
+                    .valueColumns(projectionSchema.value())
+                    .build()
+                ))
+                    .nextKsqlColAlias()
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:805]: done.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java [hunk 15]
```diff
@@ -105,13 +111,20 @@ public AggregateNode(
...
         .map(se -> SelectExpression.of(
             se.getAlias(),
             ExpressionTreeRewriter
                 .rewriteWith(aggregateExpressionRewriter::process, se.getExpression())
         ))
         .collect(Collectors.toList()));
+
+    final Set<Expression> groupings = ImmutableSet.copyOf(groupBy.getGroupingExpressions());
+    this.finalSelectExpressions = ImmutableList.copyOf(projection.stream()
+        .filter(e -> !anyKeyName || !groupings.contains(e.getExpression()))
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java:125]: Meh, that's temporary code until 'any key name' is enabled.  So... no :p

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java [hunk 16]
```diff
@@ -177,6 +190,19 @@ public PlanNode getSource() {
...
+  public void validateKeyPresent(final SourceName sinkName) {
+    final List<Expression> missing = new ArrayList<>(groupBy.getGroupingExpressions());
+
+    projection.stream()
+        .map(SelectExpression::getExpression)
+        .forEach(missing::remove);
+
+    if (!missing.isEmpty()) {
+      throwKeysNotIncludedError(sinkName, "grouping expression", missing, and());
+    }
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java:203]: The SQL you quote is explaining what would happen at the moment, not what we want to happen. 

The KLIP says 'all key columns must be in the projection', which for multi-expression aggregations is each grouping expression. 

So the check is correct, even if the KLIP may not be 100% clear on this. (I'll update the KLIP)

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 17]
```diff
@@ -112,15 +145,59 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
...
+    return Streams.concat(Stream.of(syntheticKey.name()), names);
+  }
+
+  public Expression resolveSelect(final int idx, final Expression expression) {
+    return joinKey.resolveSelect(expression, getSchema());
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  @Override
+  void validateKeyPresent(final SourceName sinkName, final Projection projection) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:174]: Yeah yeah. It's on my roadmap :p

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 18]
```diff
@@ -112,15 +145,59 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
...
+
+    final boolean atLeastOneKey = Streams.concat(
+        joinKey.getAvailableKeys().stream(),
+        joinKey.getOriginalAvailableKeys().stream()
+    ).anyMatch(projection::containsExpression);
+
+    if (!atLeastOneKey) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:181]: With this PR, yes, it defaults the key column to the left join column, though you're right in that it could be the right join column.

In the future, both key columns should be in the key.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 19]
```diff
@@ -382,9 +467,206 @@ private static DataSourceType calculateSinkType(
...
+
+    /**
+     * Called to potentially resolve a {@link JoinKeyUdf} expression into the synthetic column
+     * name.
+     *
+     * @param expression the expression to resolve.
+     * @param schema the joined schema.
+     * @return the resolved expression.
+     */
+    default Expression resolveSelect(
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:529]: Good call. (There were previously more)

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 20]
```diff
@@ -382,9 +467,206 @@ private static DataSourceType calculateSinkType(
...
+          .columnAliasGenerator(Stream.of(left, right))
+          .nextKsqlColAlias();
+    }
+
+    @Override
+    public Expression resolveSelect(
+        final Expression expression,
+        final LogicalSchema schema
+    ) {
+      return joinKeyUdf.equals(expression)
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:657]: Yeah, we currently require it to be `JOINKEY(A.id, C.foo)`, and fail if its not.  I'd of preferred to not require _any_ parameters, but then figuring out types becomes tricky without a lot more magic.

There's already a test for this... let me find it....  

Humm.. so there is a test in `joinkey.json` for the wrong params: `with wrong join key parameters`, but not a the multi-join version.

Added `unsupported joinkey parameters` to `multi-joins.json` to cover this.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java]: Done.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/StreamStreamJoin.java [hunk 21]
```diff
@@ -14,45 +14,86 @@
...
   private final ExecutionStep<KStreamHolder<K>> leftSource;
   private final ExecutionStep<KStreamHolder<K>> rightSource;
   private final Duration beforeMillis;
   private final Duration afterMillis;
 
-  public StreamStreamJoin(
+  @SuppressWarnings("unused") // Invoked by reflection
+  @JsonCreator
+  @Deprecated() // Can be removed at next incompatible version
+  private StreamStreamJoin(
```
Comment (agavra) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/StreamStreamJoin.java:50]: can jackson discover private constructors? clearly the tests pass, but I'm surprised

Comment (big-andy-coates) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/StreamStreamJoin.java:50]: Yeah, Jackson can access private constructors, methods and even fields directly.
